src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java

```diff
@@ -311,12 +314,27 @@ public Pod build() {
 
             KubernetesCloud cloud = slave.getKubernetesCloud();
 
-            String url = cloud.getJenkinsUrlOrDie();
-
-            env.put("JENKINS_URL", url);
             if (!StringUtils.isBlank(cloud.getJenkinsTunnel())) {
                 env.put("JENKINS_TUNNEL", cloud.getJenkinsTunnel());
             }
+            
+            if (!cloud.isDirectConnection()) {
+                env.put("JENKINS_URL", cloud.getJenkinsUrlOrDie());
+            } else {
+                Jenkins jenkins = Jenkins.getInstanceOrNull();
+                if(jenkins == null) throw new IllegalStateException("Jenkins instance null");
+                String host;
+                try {
+                    host = new URL(jenkins.getRootUrl()).getHost();
+                } catch (MalformedURLException e) {
+                    throw new IllegalStateException("Could not get Jenkins host name", e);
+                }
+                int port = jenkins.getTcpSlaveAgentListener().getPort();
```

Comment:
```suggestion
                int port = jenkins.getTcpSlaveAgentListener().getAdvertisedPort();
```

? (See `TcpSlaveAgentListener/index.jelly`)

Reply:
done